addOns/zest/src/main/java/org/zaproxy/zap/extension/zest/ExtensionZest.java

```diff
@@ -545,9 +554,18 @@ public ScriptNode add(ZestScriptWrapper script, boolean display) {
     }
 
     public void display(ZestScriptWrapper script, ScriptNode node, boolean expand) {
+        display(script, node, expand, true);
+    }
+
+    public void display(
+            ZestScriptWrapper script, ScriptNode node, boolean expand, boolean allowFocus) {
         if (View.isInitialised() && this.getExtScript().getScriptUI() != null) {
-            this.getExtScript().getScriptUI().selectNode(node, expand);
-            this.getExtScript().getScriptUI().displayScript(script);
+            if (allowFocus) {
+                // This function internally calls set script, which is not trivial to change at the
+                // moment
+                this.getExtScript().getScriptUI().selectNode(node, expand);
+            }
```

Comment:
Had to do it this way because this is not trivial to change. This function calls set script function which sets the focus.
Here is the stack trace for your reference:
```
java.lang.Exception
	at org.zaproxy.zap.extension.scripts.ConsolePanel.setScript(ConsolePanel.java:479)
	at org.zaproxy.zap.extension.scripts.ExtensionScriptsUI.displayScript(ExtensionScriptsUI.java:423)
	at org.zaproxy.zap.extension.scripts.ExtensionScriptsUI.displayScript(ExtensionScriptsUI.java:392)
	at org.zaproxy.zap.extension.scripts.ScriptsListPanel.selectionChanged(ScriptsListPanel.java:657)
	at org.zaproxy.zap.extension.scripts.ScriptsListPanel.lambda$getTree$4(ScriptsListPanel.java:640)
	at java.desktop/javax.swing.JTree.fireValueChanged(JTree.java:2967)
	at java.desktop/javax.swing.JTree$TreeSelectionRedirector.valueChanged(JTree.java:3456)
	at java.desktop/javax.swing.tree.DefaultTreeSelectionModel.fireValueChanged(DefaultTreeSelectionModel.java:641)
	at java.desktop/javax.swing.tree.DefaultTreeSelectionModel.notifyPathChange(DefaultTreeSelectionModel.java:1111)
	at java.desktop/javax.swing.tree.DefaultTreeSelectionModel.setSelectionPaths(DefaultTreeSelectionModel.java:297)
	at java.desktop/javax.swing.tree.DefaultTreeSelectionModel.setSelectionPath(DefaultTreeSelectionModel.java:191)
	at java.desktop/javax.swing.JTree.setSelectionPath(JTree.java:1656)
	at org.zaproxy.zap.extension.scripts.ScriptsListPanel.showInTree(ScriptsListPanel.java:714)
	at org.zaproxy.zap.extension.scripts.ExtensionScriptsUI.selectNode(ExtensionScriptsUI.java:802)
	at org.zaproxy.zap.extension.zest.ExtensionZest.display(ExtensionZest.java:568)
	at org.zaproxy.zap.extension.zest.ExtensionZest.add(ExtensionZest.java:550)
	at org.zaproxy.zap.extension.zest.dialogs.ZestRecordScriptDialog.save(ZestRecordScriptDialog.java:277)
	at org.zaproxy.zap.view.StandardFieldsDialog.savePressed(StandardFieldsDialog.java:428)
	at org.zaproxy.zap.view.StandardFieldsDialog$3.actionPerformed(StandardFieldsDialog.java:411)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6635)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6400)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5011)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
```

Reply:
Should we change core like we did for the other method?

Reply:
Yes we will need to. Will start work on this.

Reply:
at java.desktop/javax.swing.JTree.setSelectionPath(JTree.java:1656)

how will I pass the allowFocus variable to this function? I don't know if we can change this?

Reply:
alternatively what I am thinking is we can have a shouldFocus variable in ExtensionScriptUI which can be set and unset by the user and can be used to changed focus. And displayScript function will use it internally instead of using allowFocus

Reply:
Is that method requesting focus though?

Reply:
I see what are you asking, let me check.

Reply:
Let me know if you have better suggestions for this.

Reply:
just follow the stack trace you will know why we need to pass it there.

Reply:
I'm not seeing that code being called while recording, and from the stack trace it's being done on save of the script.

Reply:
Yes but even if it is called one time from there it keeps on grabbing focus I don't know why

Reply:
Can you share a screen recoding with that happening? (e.g. in slack.)

Reply:
Yes sure

Reply:
Do have any idea how can we solve this problem?

Reply:
In `selectNode` which one of the statements cause the problem `showInTree` or `setTabFocus` ?

Reply:
`showInTree`

Reply:
`getTree().setSelectionPath(tp);`? So part of the above stacktrace?

Reply:
yess

Reply:
maybe be can set the calling of showInTree as optional using allowFocus?

Reply:
Yeah, I was thinking something like that, e.g.:
```diff
--- a/addOns/scripts/src/main/java/org/zaproxy/zap/extension/scripts/ScriptsListPanel.java
+++ b/addOns/scripts/src/main/java/org/zaproxy/zap/extension/scripts/ScriptsListPanel.java
@@ -653,7 +653,7 @@ public class ScriptsListPanel extends AbstractPanel {
                         if (node.isTemplate()) {
                             extension.displayTemplate((ScriptWrapper) node.getUserObject());
                         } else {
-                            extension.displayScript((ScriptWrapper) node.getUserObject());
+                            extension.displayScript((ScriptWrapper) node.getUserObject(), allowFocus);
                         }
                     }
                     break;
@@ -689,11 +689,13 @@ public class ScriptsListPanel extends AbstractPanel {
         this.showInTree(node, false);
     }
 
-    public void showInTree(ScriptNode node, boolean expand) {
+    public void showInTree(ScriptNode node, boolean expand, boolean allowFocus) {
         TreeNode[] path = node.getPath();
         TreePath tp = new TreePath(path);
         getTree().setExpandsSelectedPaths(true);
+        this.allowFocus = allowFocus;
         getTree().setSelectionPath(tp);
+        this.allowFocus = true;
         getTree().scrollPathToVisible(tp);
         if (expand) {
             getTree().expandPath(tp);
```

Reply:
okay will do it this way